wpcom-vip-two-factor/sms-provider.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+// Bail if we don't have Twilio credentials
+if ( ! defined( 'TWILIO_SID' ) || ! defined( 'TWILIO_SECRET' ) ) {
+	return;
+}
+
+/**
+ * Class for creating an sms provider.
+ *
+ * @package Two_Factor
+ */
+class Two_Factor_SMS extends Two_Factor_Provider {
+
+	/**
+	 * The user meta token key.
+	 *
+	 * @type string
+	 */
+	const TOKEN_META_KEY = '_two_factor_sms_token';
+
+	static function get_instance() {
+		static $instance;
+		$class = __CLASS__;
+		if ( ! is_a( $instance, $class ) ) {
+			$instance = new $class;
+		}
+		return $instance;
+	}
+
+	protected function __construct() {
+		add_action( 'two-factor-user-options-' . __CLASS__, array( $this, 'user_options' ) );
+		add_filter( 'user_contactmethods', array( $this, 'user_contactmethods' ) );
+		return parent::__construct();
+	}
+
+	public function user_contactmethods( $methods ) {
+		$methods['phone'] = _x( 'Phone', 'Phone Label', 'two-factor' );
+		return $methods;
+	}
+
+	/**
+	 * Returns the name of the provider.
+	 */
+	public function get_label() {
+		return _x( 'SMS', 'Provider Label', 'two-factor' );
+	}
+
+	/**
+	 * Generate the user token.
+	 *
+	 * @param int $user_id User ID.
+	 * @return string
+	 */
+	public function generate_token( $user_id ) {
+		$token = $this->get_code();
+		update_user_meta( $user_id, self::TOKEN_META_KEY, wp_hash( $token ) );
+		return $token;
+	}
+
+	/**
+	 * Validate the user token.
+	 *
+	 * @param int    $user_id User ID.
+	 * @param string $token User token.
+	 * @return boolean
+	 */
+	public function validate_token( $user_id, $token ) {
+		$hashed_token = get_user_meta( $user_id, self::TOKEN_META_KEY, true );
+		$correct = wp_hash( $token );
+		if ( ! hash_equals( $hashed_token, $correct ) ) {
+			return false;
+		}
+		$this->delete_token( $user_id );
+		return true;
+	}
+
+	/**
+	 * Delete the user token.
+	 *
+	 * @param int $user_id User ID.
+	 */
+	public function delete_token( $user_id ) {
+		delete_user_meta( $user_id, self::TOKEN_META_KEY );
+	}
+
+	/**
+	 * Generate and send the user token.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function generate_and_send_token( $user ) {
+		require_once( WPMU_PLUGIN_DIR . '/lib/sms.php' );
+		$code = $this->generate_token( $user->ID );
+		return \Automattic\VIP\SMS\send_sms( $user->phone, $code );
+	}
+
+	/**
+	 * Prints the form that prompts the user to authenticate.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function authentication_page( $user ) {
+		if ( ! $user ) {
+			return;
+		}
+
+		if ( ! isset( $_GET['action'] ) || 'validate_2fa' !== $_GET['action'] ) {
+			$this->generate_and_send_token( $user );
+		}
+
+		// Including template.php for submit_button()
+		require_once( ABSPATH .  '/wp-admin/includes/template.php' );
```

Comment:
Sure, though it's also used in all the "core" providers:

https://github.com/georgestephanis/two-factor/blob/master/providers/class.two-factor-totp.php#L323
https://github.com/georgestephanis/two-factor/blob/master/providers/class.two-factor-fido-u2f.php#L118
https://github.com/georgestephanis/two-factor/blob/master/providers/class.two-factor-email.php#L124